mopidy_tidal/library.py

```diff
@@ -157,104 +151,111 @@ def search(self, query=None, uris=None, exact=False):
-                        artist_id = parts[2]
-                        img_uri = self.lru_artist_img.hit(artist_id)
-                        if img_uri is None:
-                            artist = session.get_artist(artist_id)
-                            img_uri = artist.image
-                            self.lru_artist_img[artist.id] = img_uri
-
-                        uri_images = [Image(uri=img_uri, width=512, height=512)]
-                    elif parts[1] == 'album':
-                        album_id = parts[2]
-                        img_uri = self.lru_album_img.hit(album_id)
-                        if img_uri is None:
-                            album = session.get_album(album_id)
-                            img_uri = album.image
-                            self.lru_album_img[album_id] = img_uri
-
-                        uri_images = [Image(uri=img_uri, width=512, height=512)]
-                    elif parts[1] == 'playlist':
-                        playlist_id = parts[2]
-                        img_uri = self.lru_playlist_img.hit(playlist_id)
-                        if img_uri is None:
-                            playlist = session.get_playlist(playlist_id)
-                            img_uri = playlist.image
-                            self.lru_playlist_img[playlist_id] = img_uri
-
-                        uri_images = [Image(uri=img_uri, width=512, height=512)]
-                    elif parts[1] == 'track':
-                        album_id = parts[3]
-                        img_uri = self.lru_album_img.hit(album_id)
-                        if img_uri is None:
-                            album = session.get_album(album_id)
-                            img_uri = album.image
-                            self.lru_album_img[album_id] = img_uri
-
-                        uri_images = [Image(uri=img_uri, width=512, height=512)]
-                        pass
-
-                images[uri] = uri_images or ()
-            except AttributeError:
-                logger.error("AttributeError when processing URI %r" % uri)
-            except HTTPError as err:
-                logger.error("HTTPError when processing URI %r" % uri)
+                images[uri] = self._get_image(uri)
+            except (AssertionError, AttributeError, HTTPError) as err:
+                logger.error(
+                    "%s when processing URI %r: %s",
+                    type(err), uri, err)
 
+        self._image_cache.update(images)
         return images
 
     def lookup(self, uris=None):
         logger.info("Lookup uris %r", uris)
-        session = self.backend._session
         if isinstance(uris, str):
             uris = [uris]
         if not hasattr(uris, '__iter__'):
             uris = [uris]
 
         tracks = []
-        for uri in uris:
+        for uri in (uris or []):
+            data = []
             try:
                 parts = uri.split(':')
-                if uri.startswith('tidal:track:'):
```

Comment:
I understand the logic behind defining cache, lookup functions depending on the URI type but I think it is less readable that way. I prefer an if/else statement, similar to what we used previously:
```python
for uri in (uris or []):
data = []
try:
parts = uri.split(':')
if uri.startswith('tidal:track:'):
    cache = self._track_cache
    lookup = self._lookup_track
elif uri.startswith('tidal:album'):
    cache = self._album_cache
    lookup = self._lookup_album
elif uri.startswith('tidal:artist'):
    cache = self._track_cache
    lookup = self._lookup_artist
elif uri.startswith('tidal:playlist'):
    cache = self._playlist_cache
    lookup = self._lookup_playlist

try:
    data = cache[uri]
except KeyError:
    data = lookup(self._session, parts)
```

Reply:
It seems some URI types fall outside the above categories. This will cause the subsequent cache lookup to fail. For now, lets stick to your refactored version :)